test/libsolidity/ASTPropertyTest.cpp

```diff
@@ -0,0 +1,224 @@
+                    advance();
+                }
+
+                skipWhitespace();
+                soltestAssert(m_currentChar == ':');
+                advance();
+                skipWhitespace();
+
+                testValue.clear();
+                while(isValidValue(m_currentChar))
+                {
+                    testValue += m_currentChar;
+                    advance();
+                }
+                parsedPairs.emplace_back(testId, testValue);
+            }
+        }
+        return parsedPairs;
+    }
+}
+
+ASTPropertyTest::ASTPropertyTest(string const& _filename):
+	TestCase(_filename)
+{
+	if (!boost::algorithm::ends_with(_filename, ".sol"))
+		BOOST_THROW_EXCEPTION(runtime_error("Invalid test contract file name: \"" + _filename + "\"."));
+
+    m_source = m_reader.source();
+    readExpectations();
+}
+
+void ASTPropertyTest::generateTestCaseValues(string& _values, bool _obtained)
+{
+    _values.clear();
+    for (auto testId: m_expectationsSequence)
+    {
+        soltestAssert(m_testCases.count(testId) > 0);
+        _values +=
+            testId +
+            ": " +
+            (_obtained ? m_testCases[testId].obtainedValue : m_testCases[testId].expectedValue)
+            + "\n";
+    }
+}
+
+void ASTPropertyTest::readExpectations()
+{
+    KeyValueParser parser{m_reader.simpleExpectations()};
+    for (auto [testId, testExpectation] : parser.parse())
+    {
+        m_testCases.emplace(testId, ASTPropertyTestCase{testId, "", testExpectation, ""});
+        m_expectationsSequence.push_back(testId);
+    }
+    generateTestCaseValues(m_expectation, false);
+}
+
+optional<Json::Value> ASTPropertyTest::findNode(Json::Value const& _root, string_view const& _property)
+{
+    if (!_property.empty())
+    {
+        string subNode = string(_property.substr(0, _property.find_first_of('.')));
+        if (subNode != _property)
+            return findNode(_root[subNode], _property.substr(_property.find_first_of('.') + 1));
+        else if (_root.isMember(subNode))
+            return make_optional(_root[subNode]);
+    }
+    return {};
+}
+
+void ASTPropertyTest::readTestedProperties(Json::Value const& _astJson)
+{
+    queue<Json::Value> nodesToVisit;
+    nodesToVisit.push(_astJson);
+    string documentation = "documentation";
+    string testCaseLine;
+
+    while(!nodesToVisit.empty())
+    {
+        auto& node = nodesToVisit.front();
+        if (node.isObject())
+        {
+            for (auto memberName: node.getMemberNames())
+                if (memberName == documentation)
+                {
+                    auto docNode = node[documentation];
+                    testCaseLine = docNode.isObject() ?
+                        docNode["text"].asString() :
+                        docNode.asString();
+
+                    soltestAssert(!testCaseLine.empty());
+                    KeyValueParser parser{testCaseLine};
+                    auto [testId, testedProperty] = parser.parse()[0];
```

Comment:
Asserting that whatever `parser.parse()` returns is not `empty()` would be preferable to a segfault.